crates/core/tedge_actors/src/test_helpers.rs

```diff
@@ -7,130 +7,307 @@ use crate::MessageSource;
+    /// let mut sender = sender_builder.build();
+    /// let receiver: SimpleMessageBox<MyMessage,NoMessage> = receiver_builder.build();
+    ///
+    /// use tedge_actors::test_helpers::MessageReceiverExt;
+    /// let mut receiver = receiver.with_timeout(Duration::from_millis(100));
+    ///
+    /// sender.send(MyMessage::Foo(1)).await?;
+    /// sender.send(MyMessage::Bar(2)).await?;
+    /// sender.send(MyMessage::Foo(3)).await?;
+    ///
+    /// receiver.assert_received_matching(
+    ///     |pat:&u32,msg:&MyMessage| msg.count() == *pat,
+    ///     [3,2],
+    /// ).await;
+    ///
+    /// # Ok(())
+    /// # }
+    ///
+    /// ```
+    async fn assert_received_matching<T, F>(&mut self, matching: F, expected: T)
+    where
+        T: IntoIterator + Send,
+        F: Fn(&T::Item, &M) -> bool,
+        F: Send,
+        T::Item: Debug + Send;
+}
+
+#[async_trait]
+impl<T, M> MessageReceiverExt<M> for T
+where
+    T: ReceiveMessages<M> + Send + Sync + 'static,
+    M: Message + Eq + PartialEq,
 {
-    for expected_msg in expected.into_iter().map(|msg| msg.into()) {
-        let actual_msg = tokio::time::timeout(timeout, messages.recv()).await;
-        assert_eq!(actual_msg, Ok(Some(expected_msg)));
+    fn with_timeout(self, timeout: Duration) -> TimedBoxReceiver<M> {
+        TimedBoxReceiver {
+            timeout,
+            inner: Box::new(self),
+        }
+    }
+
+    async fn skip(&mut self, count: usize) {
+        for _ in 0..count {
+            let _ = self.recv().await;
+        }
+    }
+
+    #[allow(clippy::needless_collect)] // To avoid issues with Send constraints
+    async fn assert_received<Samples>(&mut self, expected: Samples)
+    where
+        Samples: IntoIterator + Send,
+        M: From<Samples::Item>,
+    {
+        let expected: Vec<M> = expected.into_iter().map(|msg| msg.into()).collect();
+        for expected_msg in expected.into_iter() {
+            let actual_msg = self.recv().await;
+            assert_eq!(actual_msg, Some(expected_msg));
+        }
+    }
+
+    async fn assert_received_unordered<Samples>(&mut self, expected: Samples)
+    where
+        Samples: IntoIterator + Send,
+        M: From<Samples::Item>,
+    {
+        let expected: Vec<M> = expected.into_iter().map(|msg| msg.into()).collect();
+        self.assert_received_matching(|pat: &M, msg: &M| pat == msg, expected)
+            .await
+    }
+
+    async fn assert_received_matching<Samples, F>(&mut self, matching: F, expected: Samples)
+    where
+        Samples: IntoIterator + Send,
+        F: Fn(&Samples::Item, &M) -> bool,
+        F: Send,
+        Samples::Item: Debug + Send,
+    {
+        let mut expected: Vec<Samples::Item> = expected.into_iter().collect();
+        let mut received = Vec::new();
+
+        while let Some(msg) = self.recv().await {
+            expected.retain(|pat| !matching(pat, &msg));
```

Comment:
Not a request for change but more of a personal comment: The `HashSet` impl was far simpler than this, especially with that `!matching`. I always find negation based logic difficult to follow. But I agree that the `Hash` constraint was a bigger limitation. So, this is fine.

Reply:
Yes, enforcing a `Hash` constraint was a pain in practice.

The main concern I have with `retain` is performance, leading to a received-all check with O(n^2) complexity. But this is not a problem for tests with less than 100 samples.